File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/listener/condition/EntityCustomNameContains.java
package me.lokka30.levelledmobs.bukkit.listener.condition;

import java.util.Collections;
import me.lokka30.levelledmobs.bukkit.logic.context.Context;
import me.lokka30.levelledmobs.bukkit.logic.function.process.Process;
import me.lokka30.levelledmobs.bukkit.logic.function.process.condition.Condition;
import me.lokka30.levelledmobs.bukkit.util.Log;
import me.lokka30.levelledmobs.bukkit.util.modal.ModalCollection.Type;
import me.lokka30.levelledmobs.bukkit.util.modal.ModalList;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;

public class EntityCustomNameContains extends Condition {

    /* vars */

    /*
    Modal list of entity types

    Do not make this variable `final`.
     */
    private ModalList<String> modalList;

    /* constructors */

    public EntityCustomNameContains(Process process, final CommentedConfigurationNode node) {
        super(process, node);

        try {
            if (getConditionNode().hasChild("in-list")) {
                this.modalList = new ModalList<>(
                    getConditionNode().node("in-list")
                        .getList(String.class, Collections.emptyList()),
                    Type.INCLUSIVE
                );
            } else if (getConditionNode().hasChild("not-in-list")) {
                this.modalList = new ModalList<>(
                    getConditionNode().node("not-in-list")
                        .getList(String.class, Collections.emptyList()),
                    Type.EXCLUSIVE
                );
            } else {
                //TODO make better error message
                Log.sev("entity custom name contains condition error: no in-list/not-in-list declaration",
                    true);
            }
        } catch (ConfigurateException ex) {
            //TODO make better error message
            Log.sev("entity custom name contains condition error: unable to parse yml", true);
        }
    }

    /* methods */

    @Override
    public boolean applies(Context context) {

        assert context.getEntity() != null;

        final var customName = context.getEntity().getCustomName();

        if(customName == null)
            return getModalList().getType() == Type.EXCLUSIVE;

        final var hasAsterisk = getModalList().getItems().stream().anyMatch("*"::contains);

        if(hasAsterisk) {
            return switch(getModalList().getType()) {
                case INCLUSIVE -> true;
                case EXCLUSIVE -> false;
            };
        }

        final var contains = getModalList().getItems().stream()
            .anyMatch(customName::contains);

        return switch(getModalList().getType()) {
            case INCLUSIVE -> contains;
            case EXCLUSIVE -> !contains;
        };
    }


    /* getters and setters */

    @NotNull
    public ModalList<String> getModalList() {
        return modalList;
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/condition/impl/EntityCustomNameContains.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.condition.impl

import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.Process
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.condition.Condition
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.sev
import io.github.arcaneplugins.levelledmobs.bukkit.util.modal.ModalCollection
import io.github.arcaneplugins.levelledmobs.bukkit.util.modal.ModalList
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.ConfigurateException

class EntityCustomNameContains(
    process: Process,
    node: CommentedConfigurationNode
): Condition(process, node) {
    var modalList: ModalList<String>? = null

    init {
        try {
            if (conditionNode.hasChild("in-list")) {
                modalList = ModalList(
                    conditionNode.node("in-list")
                        .getList(String::class.java, emptyList()),
                    ModalCollection.Mode.INCLUSIVE
                )
            } else if (conditionNode.hasChild("not-in-list")) {
                modalList = ModalList(
                    conditionNode.node("not-in-list")
                        .getList(String::class.java, emptyList()),
                    ModalCollection.Mode.EXCLUSIVE
                )
            } else {
                //TODO make better error message
                sev(
                    "entity custom name contains condition error: no in-list/not-in-list declaration",
                    true
                )
            }
        } catch (ex: ConfigurateException) {
            //TODO make better error message
            sev("entity custom name contains condition error: unable to parse yml", true)
        }
    }

    override fun applies(context: Context): Boolean {
        requireNotNull(context.entity) { "Entity was not be null" }
        val customName = context.entity!!.customName ?:
            return modalList!!.mode == ModalCollection.Mode.EXCLUSIVE

        val hasAsterisk: Boolean = modalList!!.items.stream().anyMatch { s: String? ->
            "*".contains(s!!)
        }

        if (hasAsterisk) {
            return when (modalList!!.mode) {
                ModalCollection.Mode.INCLUSIVE -> true
                ModalCollection.Mode.EXCLUSIVE -> false
            }
        }

        val contains: Boolean = modalList!!.items.stream()
            .anyMatch { s: String? -> customName.contains(s!!) }

        return when (modalList!!.mode) {
            ModalCollection.Mode.INCLUSIVE -> contains
            ModalCollection.Mode.EXCLUSIVE -> !contains
        }
    }
}